File: lollys-lotto-rust-sdk/src/instructions/start_lotto_game.rs
use crate::instructions::*;

pub fn start_lotto_game(
    round: u64,
    ticket_price: u64,
    game_duration: u64,
    round_name: String,
    authority: &Pubkey,
    lollys_lotto: &Pubkey,
    lotto_game: &Pubkey,
    lotto_game_vault_signer: &Pubkey,
    lotto_game_vault: &Pubkey,
    lotto_game_mint: &Pubkey,
    event_emitter: &Pubkey,
) -> Instruction {
    let data = lollys_lotto::instruction::StartLottoGame{
        round,
        ticket_price,
        game_duration,
        round_name,
    }.data();

    let accounts = lollys_lotto::accounts::StartLottoGame{
        authority: *authority,
        lollys_lotto: *lollys_lotto,
        lotto_game: *lotto_game,
        lotto_game_vault_signer: *lotto_game_vault_signer,
        lotto_game_vault: *lotto_game_vault,
        lotto_game_mint: *lotto_game_mint,
        event_emitter: *event_emitter,
        token_program: token::ID,
        associated_token_program: associated_token::ID,
        system_program: system_program::ID,
    }.to_account_metas(None);
    Instruction {
        program_id: lollys_lotto::ID,
        accounts,
        data,
    }
}

File: lollys-lotto-ts-sdk/src/codegen/instructions/startLottoGame.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface StartLottoGameArgs {
  round: BN
  ticketPrice: BN
  gameDuration: BN
  roundName: string
}

export interface StartLottoGameAccounts {
  authority: PublicKey
  lollysLotto: PublicKey
  lottoGame: PublicKey
  lottoGameVaultSigner: PublicKey
  /**
   * This instruction initializes the token account
   * required for storing this LottoGame's ticket amount collections in USDC.
   */
  lottoGameVault: PublicKey
  /** Needed for account initialization */
  lottoGameMint: PublicKey
  eventEmitter: PublicKey
  tokenProgram: PublicKey
  associatedTokenProgram: PublicKey
  systemProgram: PublicKey
}

export const layout = borsh.struct([
  borsh.u64("round"),
  borsh.u64("ticketPrice"),
  borsh.u64("gameDuration"),
  borsh.str("roundName"),
])

export function startLottoGame(
  args: StartLottoGameArgs,
  accounts: StartLottoGameAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.authority, isSigner: true, isWritable: true },
    { pubkey: accounts.lollysLotto, isSigner: false, isWritable: true },
    { pubkey: accounts.lottoGame, isSigner: false, isWritable: true },
    {
      pubkey: accounts.lottoGameVaultSigner,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.lottoGameVault, isSigner: false, isWritable: true },
    { pubkey: accounts.lottoGameMint, isSigner: false, isWritable: false },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    {
      pubkey: accounts.associatedTokenProgram,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([170, 200, 70, 202, 157, 229, 45, 144])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      round: args.round,
      ticketPrice: args.ticketPrice,
      gameDuration: args.gameDuration,
      roundName: args.roundName,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}
